### PR TITLE
Package organization; license; yaml annotations functions.

### PR DESCRIPTION
**Pull request by kkdang**

### CODE REVIEW

File: pythonSynapseUtils/annotationsYaml.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
+import yaml
+
+
+def checkAgainstDict(annotDict, project, syn){
+	''' Compares annotations in use in a project against those specified by a dictionary and prints the ones that do not match the dictionary.'''
```
Comment (kkdang) [inline pythonSynapseUtils/annotationsYaml.py]: Make PEP standard format.

File: pythonSynapseUtils/annotationsYaml.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	yamlEnt = syn.get(annotDict)
+	annotations = yaml.load(file(yamlEnt.path))
+
+	allKeysInProject = set()
+	allValsInProject = set()
+
+	systemKeysToExclude = ['creationDate', 'etag', 'id', 'uri', 'accessControl']
```
Comment (kkdang) [inline pythonSynapseUtils/annotationsYaml.py]: Consider moving to module level.

File: pythonSynapseUtils/annotationsYaml.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	results = syn.chunkedQuery('select id from file where projectId=="%s"' % project)
+	for result in results:
+		temp = syn.getAnnotations(result['file.id'])
+		for key in temp:
+			if key in systemKeysToExclude: continue
+			if not key in allKeysInProject:
+				allKeysInProject.add(key)
+			if isinstance(temp[key], list):
+				for val in temp[key]:
+					if str(val) in allValsInProject: continue
```
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: Don't check if in the set (set does that)

File: pythonSynapseUtils/annotationsYaml.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+					allValsInProject.add(str(val))
+			else:
+				if str(temp[key]) in allValsInProject: continue
```
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: Don't check if in the set (set does that)

File: pythonSynapseUtils/annotationsYaml.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+	systemKeysToExclude = ['creationDate', 'etag', 'id', 'uri', 'accessControl']
+
+	results = syn.chunkedQuery('select id from file where projectId=="%s"' % project)
+	for result in results:
+		temp = syn.getAnnotations(result['file.id'])
+		for key in temp:
+			if key in systemKeysToExclude: continue
+			if not key in allKeysInProject:
+				allKeysInProject.add(key)
```
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: Don't check if in the set (set does that)

File: pythonSynapseUtils/annotationsYaml.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+
+	allKeysInVocab = set(annotations.keys())
+	if not allKeysInProject <= allKeysInVocab:
+		print 'Keys in use that are not found in dictionary: '
+		for item in allKeysInProject.difference(allKeysInVocab):
+			print '%s' % item
+
+
+	allValsInVocab = list()
```
Comment (xindiguo) [inline pythonSynapseUtils/annotationsYaml.py]: Use set instead of list

File: pythonSynapseUtils/annotationsYaml.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	return count
+
+
+def countPerAnnot(annotDict, project, syn, grouping=None){
+	'''Counts number of items annotated to each key-value pair in the given dictionary within the given project.'''
+
+	yamlEnt = syn.get(annotDict)
+	annotations = yaml.load(file(yamlEnt.path))
+
+	if grouping is not None: # This block counts annotation terms stratified by one term as specified in argument 'grouping'
```
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: pandas....

File: pythonSynapseUtils/annotationsYaml.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
+import yaml
+
+
+def checkAgainstDict(annotDict, project, syn){
+	''' Compares annotations in use in a project against those specified by a dictionary and prints the ones that do not match the dictionary.'''
+	yamlEnt = syn.get(annotDict)
+	annotations = yaml.load(file(yamlEnt.path))
```
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: Abstract this part - specify the structure of annotations, and allow user to get it in (then my tables-based version could use this as well)

File: pythonSynapseUtils/annotationsYaml.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
+import yaml
+
+
+def checkAgainstDict(annotDict, project, syn){
+	''' Compares annotations in use in a project against those specified by a dictionary and prints the ones that do not match the dictionary.'''
+	yamlEnt = syn.get(annotDict)
```
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: confusing that annotDict is really a Synapse Entity ID. Docs will help, but explicit variable name would be good (`yamlAnnotSynId`)

File: pythonSynapseUtils/annotationsYaml.py (hunk 8 above)
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: also, an anti-pattern. Better:

```
with open(yamlEnt.path) as f:
    annotations = yaml.load(f)
```

File: pythonSynapseUtils/annotationsYaml.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+	allKeysInVocab = set(annotations.keys())
+	if not allKeysInProject <= allKeysInVocab:
+		print 'Keys in use that are not found in dictionary: '
+		for item in allKeysInProject.difference(allKeysInVocab):
+			print '%s' % item
+
+
+	allValsInVocab = list()
+	for key in annotations:
```
Comment (kdaily) [inline pythonSynapseUtils/annotationsYaml.py]: ` for key, temp in annotations.iteritems():`

File: setup.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+setup(name='pythonSynapseUtils',
+      version='0.1',
+      description='Common functions for data curation with Synapse.',
+      long_description=open('README.md').read(),
+      url='https://github.com/Sage-Bionetworks/pythonSynapseUtils',
+      author='Abhishek Pratap, Kenny Daily, Kristen Dang, Xindi Guo',
+      license='Apache',
+      packages=['pythonSynapseUtils'],
+      install_requires=['synapseclient>1.0'],
```
Comment (xindiguo) [inline setup.py]: synapseclient >1.2

> Table functionality is available with the most release of both of our analytical clients as well (R Client version 1.5-4 and Python Client version 1.2)

source: [https://sagesynapse.wordpress.com/category/synapse-features/](url)

**Pull request merged**